File: ai_agents/agents/examples/voice-assistant-realtime/main_python/agent/events.py
from pydantic import BaseModel
from typing import Literal, Optional, Union, Dict, Any
from ten_ai_base.types import LLMToolMetadata


# ==== Base Event ====

class AgentEventBase(BaseModel):
    """Base class for all agent-level events."""
    type: Literal["cmd", "data"]
    name: str


# ==== CMD Events ====

class UserJoinedEvent(AgentEventBase):
    """Event triggered when a user joins the session."""
    type: Literal["cmd"] = "cmd"
    name: Literal["on_user_joined"] = "on_user_joined"

class UserLeftEvent(AgentEventBase):
    """Event triggered when a user leaves the session."""
    type: Literal["cmd"] = "cmd"
    name: Literal["on_user_left"] = "on_user_left"

class ToolRegisterEvent(AgentEventBase):
    """Event triggered when a tool is registered by the user."""
    type: Literal["cmd"] = "cmd"
    name: Literal["tool_register"] = "tool_register"
    tool: LLMToolMetadata
    source: str


# ==== DATA Events ====

class MLLMRequestTranscriptEvent(AgentEventBase):
    """Event triggered when MLLM request transcript is received (partial or final)."""
    type: Literal["data"] = "data"
    name: Literal["mllm_request_transcript"] = "mllm_request_transcript"
    content: Optional[str] = None
    delta: Optional[str] = None
    final: bool
    metadata: Dict[str, Any]


class MLLMResponseTranscriptEvent(AgentEventBase):
    """Event triggered when LLM returns a streaming response."""
    type: Literal["llm"] = "llm"
    name: Literal["llm_response"] = "llm_response"
    delta: str
    content: str
    is_final: bool

# ==== Unified Event Union ====

AgentEvent = Union[
    UserJoinedEvent,
    UserLeftEvent,
    ToolRegisterEvent,
    MLLMRequestTranscriptEvent,
    MLLMResponseTranscriptEvent
]

File: ai_agents/agents/examples/voice-assistant-nodejs/ten_packages/extension/main_nodejs/src/agent/events.ts
//
// This file is part of TEN Framework, an open source project.
// Licensed under the Apache License, Version 2.0.
//

// ==== Base Event ====
export abstract class AgentEventBase {
    constructor(
        public type: "cmd" | "data" | "message",
        public name: string
    ) { }
}

// ==== CMD Events ====

export class UserJoinedEvent extends AgentEventBase {
    constructor() {
        super("cmd", "on_user_joined");
    }
}

export class UserLeftEvent extends AgentEventBase {
    constructor() {
        super("cmd", "on_user_left");
    }
}

export class ToolRegisterEvent extends AgentEventBase {
    constructor(
        public tool: any,          // TODO: replace with LLMToolMetadata type
        public source?: string
    ) {
        super("cmd", "tool_register");
    }
}

// ==== DATA Events ====

export class ASRResultEvent extends AgentEventBase {
    constructor(
        public text: string,
        public final: boolean,
        public metadata?: Record<string, any>
    ) {
        super("data", "asr_result");
    }
}

// ==== MESSAGE Events ====

export class LLMResponseEvent extends AgentEventBase {
    constructor(
        public delta: string,
        public content: string,
        public is_final: boolean,
        public kind: "message" | "reasoning" = "message"
    ) {
        super("data", "llm_response");
        this.content = content
        this.delta = delta
        this.is_final = is_final

    }
}

// ==== Union Type Helper ====

export type AgentEvent =
    | UserJoinedEvent
    | UserLeftEvent
    | ToolRegisterEvent
    | ASRResultEvent
    | LLMResponseEvent;
